Use BLAS for linear algebra when available

**Problem**

Currently `GF2.__matmul__` inherits from `Array.__matmul__` that through a set of checks refers to `np.ndarray.__matmul__` via [this line](https://github.com/mhostetter/galois/blob/1e377a1f836d48777e3f0afac7156d252d51a6dd/src/galois/_domains/_linalg.py#L191).

Numpy implements matrix multiplication for integer matrices through a slower algorithm than for float matrices, where BLAS is used, and there is no plan to use it for integer matrices, as per [this issue](https://github.com/numpy/numpy/issues/14556). So there is a possibility to gain speedup by just changing data type.

**Proposed change**

I think in general this is relevant for more than just `GF2` fields and can be addressed in data type changes [here](https://github.com/mhostetter/galois/blob/1e377a1f836d48777e3f0afac7156d252d51a6dd/src/galois/_domains/_linalg.py#L44)

**Obtained speedup**

This adaptation changes the time scaling, as is shown on the Figure below (x axis shows matrix dimension) that was computed for `GF2` objects. The orange curve shows multiplication as is, the green curve shows multiplication with type conversion to and from float (and blue is another option we tried).

![Image](https://github.com/user-attachments/assets/b4df9824-867c-4410-b363-1d9cf13edcc4)

**Action item**

Would you be interested in implementing this kind of change?

fp64 can represent integers exactly between -2^53 and 2^53. The example above shows inequality because it goes above 2^53.

With uint64, one can go up to 2^64.  (I can amend the code snippet above to make it fail as well for the integer case)

So I'm wondering if it's worth restricting the range from 2^64 to 2^53 to gain that 100x speedup?

```
    import numpy as np

    dim1 = 10

    for p in [2, 7, 1_009, 1_000_003, 100_000_003]:
        for dim2 in [10, 100, 1_000, 10_000, 100_000, 1_000_000]:
            A = np.random.randint(0, p, (dim1, dim2))
            B = np.random.randint(0, p, (dim2, dim1))

            AB_int = A.astype(np.int64) @ B.astype(np.int64)
            AB_float = np.round(A.astype(np.float64) @ B.astype(np.float64)).astype(np.int64)
            print(p, dim1, dim2, np.array_equal(AB_int, AB_float))

            assert np.all(AB_int >= 0), f"failed for p={p} and dim2={dim2}"
            # >> AssertionError: failed for p=100000003 and dim2=100000
```

--- a/src/galois/_domains/_linalg.py
+++ b/src/galois/_domains/_linalg.py
@@ -13,7 +13,6 @@ from numba import int64
 
 from .._helper import verify_isinstance
 from ._function import Function, FunctionMixin
-from ._meta import DTYPES
 
 if TYPE_CHECKING:
     from ._array import Array
@@ -35,12 +34,21 @@ def _lapack_linalg(field: Type[Array], a: Array, b: Array, function, out=None, n
     a = a.view(np.ndarray)
     b = b.view(np.ndarray)
 
-    # Determine the minimum dtype to hold the entire product and summation without overflowing
+    # Determine the maximum possible integer summation. Find the smallest floating-point data type that can hold that
+    # integer value. This enables the linear algebra to be computed using BLAS, which is faster than integer arithmetic.
+    # The maximum integer that can be represented in a float64 is 2**(np.finfo(np.float64).nmant + 1). If
+    # floating-point arithmetic is not possible, then np.int64 or np.object_ is used.
     if n_sum is None:
         n_sum = 1 if len(a.shape) == 0 else max(a.shape)
     max_value = n_sum * (field.characteristic - 1) ** 2
-    dtypes = [dtype for dtype in DTYPES if np.iinfo(dtype).max >= max_value]
-    dtype = np.object_ if len(dtypes) == 0 else dtypes[0]
+    if max_value <= 2 ** (np.finfo(np.float32).nmant + 1):
+        dtype = np.float32
+    elif max_value <= 2 ** (np.finfo(np.float64).nmant + 1):
+        dtype = np.float64
+    elif max_value <= np.iinfo(np.int64).max:
+        dtype = np.int64
+    else:
+        dtype = np.object_
     a = a.astype(dtype)
     b = b.astype(dtype)
 

